tests/templates/under_development/test_mapping_udev.py

```diff
@@ -52,31 +57,60 @@ class Parameters:
-    """
-    # Invoke
-    graphs: list[rdflib.Graph] = list(parameters.mapper().apply_mapping(parameters.data))
-
-    # Ensure only one graph
-    assert len(graphs) == 1
-
-    # Compare, output to file if not valid
-    if not (is_same := graph_comparer(graphs[0], parameters.expected)):
-        graphs[0].serialize(f"tests/templates/under_development/{parameters.mapper().template_id}-result.ttl")
-    assert is_same
-
-    # Check that there are no `None`s in the Graph
-    # This check is important. As some fields are optional they can be `None`
-    # at runtime. Unfortunately, `None` is valid in many contexts in Python,
-    # including string formatting. This means that type-checking is unable to
-    # determine whether a statement is valid in our specific context. As such,
-    # we check here to see if any `None`s have snuck their way into the RDF.
-    assert "None" not in graphs[0].serialize(format="ttl")
+class TestMapping:
+    def test_apply_mapping(self, parameters: Parameters, graph_comparer: Callable) -> None:
+        """Tests apply mapping for the specific template while in development.
+
+        Args:
+            parameters (Parameters): Parameters for the test.
+            graph_comparer (Callable): Graph comparison fixture.
+        """
+        # Invoke
+        graphs: list[rdflib.Graph] = list(parameters.mapper().apply_mapping(parameters.data))
+
+        # Ensure only one graph
+        assert len(graphs) == 1
+
+        # Compare, output to file if not valid
+        if not (is_same := graph_comparer(graphs[0], parameters.expected)):
+            graphs[0].serialize(f"tests/templates/under_development/{parameters.mapper().template_id}-result.ttl")
+        assert is_same
+
+        # Check that there are no `None`s in the Graph
+        # This check is important. As some fields are optional they can be `None`
+        # at runtime. Unfortunately, `None` is valid in many contexts in Python,
+        # including string formatting. This means that type-checking is unable to
+        # determine whether a statement is valid in our specific context. As such,
+        # we check here to see if any `None`s have snuck their way into the RDF.
+        assert "None" not in graphs[0].serialize(format="ttl")
+
+    def test_shapes(self, parameters: Parameters) -> None:
+        """Tests generated rdf against expected shape graph.
+
+        Args:
+            parameters (Parameters): Parameters for the test.
+        """
+        # Create data graph
+        data_graphs = (g for g in parameters.mapper().apply_mapping(parameters.data))
+
+        # Create shape graph
+        shape_graph = rdflib.Graph().parse(data=parameters.shacl)
+
+        # Perform validation
+        for data_graph in data_graphs:
+            valid, _, report = pyshacl.validate(data_graph=data_graph, shacl_graph=shape_graph)
+            if not valid:
+                raise AssertionError(report)
```

Comment:
The 2-expression form of assert does the same thing. e.g. `assert valid, report`

Reply:
oh nice never knew that!